src/main/java/seedu/address/model/ModelManager.java

```diff
@@ -79,6 +88,49 @@ public synchronized void addTask(Task person) {
         indicateTaskBookChanged();
     }
 
+    //=============for undo and redo===================================
+    @Override
+    public Command undo() throws IllegalValueException {
+
+        if (head <= 0) {
+            System.out.println(head);
+            throw new IllegalValueException("No actions to undo");
+        }
+        head--;
+        Commit commit = commits.get(head);
+        resetData(new TaskBook(commit.getTaskBook()));
+        return commit.getCommand();
+    }
+
+    @Override
+    public void recordState(Command command) {
+        commits.add(new Commit(command, new TaskBook(getTaskBook())));
+        this.head = commits.size() - 1;
```

Comment:
What happens if I do the following sequence of events?
1. `recordState()`
2. `undo()`
3. `recordState()`
4. `undo()`
5. `undo()`

I think there is a bug here. You really need to write tests for `ModelManager` to prove that the logic implemented for undo/redo does not contain any bugs.